File: src/lib/connection-presets.ts
import type { TransportSecurity, TransportSecurityMode } from '$lib/ir';

export type DatabaseLocation = 'local' | 'cloud' | 'organization';

export type SecurityLabel = {
	tone: 'success' | 'warning' | 'danger';
	text: string;
};

export function transportPresetForLocation(location: DatabaseLocation): TransportSecurity {
	if (location === 'local') {
		return {
			mode: 'Disabled',
			ca_pem: null,
			oracle_wallet_location: null,
			insecure_acknowledged: false,
			legacy_implicit: false
		};
	}

	return {
		mode: 'VerifyIdentity',
		ca_pem: null,
		oracle_wallet_location: null,
		insecure_acknowledged: false,
		legacy_implicit: false
	};
}

export function securityLabelForMode(mode: TransportSecurityMode): SecurityLabel {
	switch (mode) {
		case 'VerifyIdentity':
		case 'VerifyCa':
			return { tone: 'success', text: 'Secure connection' };
		case 'EncryptOnly':
			return { tone: 'warning', text: 'Encrypted (certificate not verified)' };
		case 'Disabled':
			return { tone: 'danger', text: 'Not encrypted - local only' };
	}
}

export function isLocalHost(host: string): boolean {
	const normalized = host.trim().toLowerCase();
	return (
		normalized === 'localhost' ||
		normalized === '127.0.0.1' ||
		normalized === '::1' ||
		normalized === '[::1]'
	);
}

export function inferLocation(host: string): Exclude<DatabaseLocation, 'organization'> {
	return isLocalHost(host) ? 'local' : 'cloud';
}

File: shared/src/main/kotlin/com/safedb/connection/ConnectionPresets.kt
package com.safedb.connection

import com.safedb.model.TransportSecurity
import com.safedb.model.TransportSecurityMode

enum class DatabaseLocation {
    Local,
    Cloud,
    Organization,
}

data class DialectInfo(val value: com.safedb.model.Dialect, val label: String, val defaultPort: Int)

val DIALECTS: List<DialectInfo> =
    listOf(
        DialectInfo(com.safedb.model.Dialect.Postgres, "PostgreSQL", 5432),
        DialectInfo(com.safedb.model.Dialect.MySql, "MySQL", 3306),
        DialectInfo(com.safedb.model.Dialect.Mssql, "SQL Server", 1433),
        DialectInfo(com.safedb.model.Dialect.Oracle, "Oracle", 1521),
    )

fun transportPresetForLocation(location: DatabaseLocation): TransportSecurity =
    if (location == DatabaseLocation.Local) {
        TransportSecurity(mode = TransportSecurityMode.Disabled)
    } else {
        TransportSecurity(mode = TransportSecurityMode.VerifyIdentity)
    }

fun isLocalHost(host: String): Boolean {
    val normalized = host.trim().lowercase()
    return normalized == "localhost" ||
        normalized == "127.0.0.1" ||
        normalized == "::1" ||
        normalized == "[::1]"
}

fun inferLocation(host: String): DatabaseLocation =
    if (isLocalHost(host)) DatabaseLocation.Local else DatabaseLocation.Cloud
